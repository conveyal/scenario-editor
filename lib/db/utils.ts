import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
