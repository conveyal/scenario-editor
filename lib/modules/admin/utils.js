import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
