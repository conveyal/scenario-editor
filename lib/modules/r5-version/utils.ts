import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
