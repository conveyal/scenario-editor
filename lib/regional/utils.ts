function getDefaultCutoff(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.cutoffsMinutes))
    return a.cutoffsMinutes[Math.floor(a.cutoffsMinutes.length / 2)]
  return a.cutoffMinutes
}

function getDefaultDestinationPointSet(a: CL.RegionalAnalysis): string | void {
  const ids = a.destinationPointSetIds
  if (Array.isArray(ids)) return ids[0]
}

function getDefaultPercentile(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.travelTimePercentiles))
    return a.travelTimePercentiles[
      Math.floor(a.travelTimePercentiles.length / 2)
    ]
  return a.travelTimePercentile
}

/**
 * Choose
 */
export function getDefaultVariants(analysis: CL.RegionalAnalysis) {
  return {
    cutoff: getDefaultCutoff(analysis),
    destinationPointSetId: getDefaultDestinationPointSet(analysis),
    percentile: getDefaultPercentile(analysis)
  }
}

export function getComparisonVariants(comparison: CL.RegionalAnalysis) {
  return {
    comparisonCutoff: getDefaultCutoff(comparison),
    comparisonDestinationPointSetId: getDefaultDestinationPointSet(comparison),
    comparisonPercentile: getDefaultPercentile(comparison)
  }
}

/**
 * Don't compare incomparable analyses.
 */
export function getValidComparisonAnalyses(
  allAnalyses: CL.RegionalAnalysis[],
  activeJobs: CL.RegionalJob[],
  analysis: CL.RegionalAnalysis
) {
  return allAnalyses.filter(
    (c) =>
      activeJobs.findIndex((job) => job.jobId === c._id) === -1 &&
      c.request.originPointSetKey == null &&
      c.zoom === analysis.zoom &&
      c.width === analysis.width &&
      c.height === analysis.height &&
      c.north === analysis.north &&
      c.west === analysis.west
  )
}
