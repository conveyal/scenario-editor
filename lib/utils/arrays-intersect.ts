// Do the arrays intersect?
export default function arraysIntersect<T>(a1: T[], a2: T[]) {
  for (const v of a1) {
    if (a2.indexOf(v) > -1) return true
  }
  return false
}
