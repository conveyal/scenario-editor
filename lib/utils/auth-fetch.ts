import {parse} from 'cookie'

import {safeFetch, SafeResponse} from './safe-fetch'

export function createAuthHeaders(user?: CL.User) {
  const headers = {
    Authorization: `bearer ${user?.idToken}`
  }

  // Add the admin access group for administrators if it exists.
  if (user?.accessGroup === process.env.NEXT_PUBLIC_ADMIN_ACCESS_GROUP) {
    const adminTempAccessGroup = parse(document.cookie).adminTempAccessGroup
    if (adminTempAccessGroup?.length > 0) {
      headers['X-Conveyal-Access-Group'] = adminTempAccessGroup
    }
  }
  return headers
}

/**
 * Fetch wrapper that includes authentication. Defaults headers to JSON.
 */
export default function authFetch<T>(
  url: string,
  user: CL.User,
  parser: (res: Response) => Promise<T> = (res) => res.json(),
  options?: RequestInit
): Promise<SafeResponse<T>> {
  return safeFetch(url, parser, {
    mode: 'cors',
    ...options,
    headers: {
      ...createAuthHeaders(user),
      ...options?.headers
    }
  })
}
