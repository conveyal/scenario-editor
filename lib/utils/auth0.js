// @flow
import message from '@conveyal/woonerf/message'
import Auth0Lock from 'auth0-lock'

const CLIENT_ID = process.env.AUTH0_CLIENT_ID
const DOMAIN = process.env.AUTH0_DOMAIN

export const authIsRequired = CLIENT_ID && DOMAIN

const localStorage = window && window.localStorage
  ? window.localStorage
  : {
    getItem () {},
    setItem () {}
  }

const lock = authIsRequired
  ? new Auth0Lock(CLIENT_ID, DOMAIN, {
    auth: {
      params: {
        scope: 'openid email profile analyst offline_access'
      },
      redirect: false
    },
    allowSignUp: false,
    theme: {
      logo: 'https://s3-eu-west-1.amazonaws.com/analyst-logos/conveyal-128x128.png',
      primaryColor: '#2389c9'
    },
    closable: false,
    languageDictionary: {
      title: message('authentication.logIn')
    }
  })
  : {
    checkSession () {},
    getUserInfo () {},
    hide () {},
    logout () {},
    on () {},
    show () {}
  }

function flashError (error) {
  console.error(error)
  lock.show({
    flashMessage: {
      type: 'error',
      text: error.error_description
    }
  })
}

// Set the default handler for auth errors
lock.on('authorization_error', flashError)
lock.on('unrecoverable_error', flashError)

export function login (next: (any) => void) {
  function handleAuthResult (authResult) {
    console.log('handleAuthResult', authResult)
    lock.getUserInfo(authResult.accessToken, (error, profile) => {
      if (error) {
        localStorage.removeItem('user')
        lock.show({
          flashMessage: {
            type: 'error',
            text: error.error_description
          }
        })
      } else {
        localStorage.setItem('user', JSON.stringify(profile))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      }
    })
  }

  lock.checkSession({}, (error, authResult) => {
    console.log('checkSession', error, authResult)
    if (error || !authResult) {
      lock.show()
      lock.on('authenticated', handleAuthResult)
    } else {
      try {
        const profile = JSON.parse(localStorage.getItem('user'))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      } catch (e) {
        console.error('error processing user profile from localStorage', e)
        handleAuthResult(authResult)
      }
    }
  })
}

export function logout () {
  localStorage.removeItem('user')
  lock.logout({returnTo: `${window.location.origin}/login`})
}
