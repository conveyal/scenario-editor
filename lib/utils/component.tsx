import {UserProvider} from '@auth0/nextjs-auth0'
import {ThemeProvider, extendTheme} from '@chakra-ui/react'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {localUser} from 'lib/user'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

const theme = extendTheme({})

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <UserProvider user={localUser}>
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <RouterContext.Provider value={router}>
            <Map>{props.children}</Map>
          </RouterContext.Provider>
        </ThemeProvider>
      </Provider>
    </UserProvider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Helper function to mount and unmount a component.
 */
export function testComponentMount(Component, props) {
  test(`Component mount(${getName(Component)})`, () => {
    const c = testComponent(Component, props)
    const t = c.mount()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
