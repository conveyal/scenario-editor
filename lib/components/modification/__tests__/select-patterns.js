import {testAndSnapshot} from 'lib/utils/component'

import SelectPatterns from '../select-patterns'

testAndSnapshot(SelectPatterns, {
  onChange: jest.fn(),
  trips: null
})
