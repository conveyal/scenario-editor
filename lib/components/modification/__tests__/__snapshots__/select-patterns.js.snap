// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component SelectPatterns snapshot(mount) 1`] = `
<SelectPatterns
  onChange={[MockFunction]}
  routePatterns={
    Array [
      Object {
        "geometry": Object {
          "coordinates": Array [
            Array [
              -122.0246,
              36.9707,
            ],
            Array [
              -122.0279,
              37.049,
            ],
            Array [
              -121.9799,
              37.2299,
            ],
            Array [
              -121.9445,
              37.324,
            ],
            Array [
              -121.936,
              37.353,
            ],
            Array [
              -121.924,
              37.365,
            ],
          ],
          "type": "LineString",
        },
        "name": "Mock Pattern",
        "pattern_id": "mock-pattern-id",
        "stops": Array [
          Object {
            "stop_id": "1",
            "stop_lat": 36.9707,
            "stop_lon": -122.0246,
            "stop_name": "Santa Cruz",
          },
          Object {
            "stop_id": "2",
            "stop_lat": 37.049,
            "stop_lon": -122.0279,
            "stop_name": "Scotts Valley",
          },
        ],
        "trips": Array [
          Object {
            "duration": 1234,
            "end_time": 13579,
            "start_time": 12345,
            "trip_headsign": "To Downtown",
            "trip_id": "abcd",
            "trip_short_name": "The Express",
          },
        ],
      },
    ]
  }
  trips={null}
>
  <FormControl
    routePatterns={
      Array [
        Object {
          "geometry": Object {
            "coordinates": Array [
              Array [
                -122.0246,
                36.9707,
              ],
              Array [
                -122.0279,
                37.049,
              ],
              Array [
                -121.9799,
                37.2299,
              ],
              Array [
                -121.9445,
                37.324,
              ],
              Array [
                -121.936,
                37.353,
              ],
              Array [
                -121.924,
                37.365,
              ],
            ],
            "type": "LineString",
          },
          "name": "Mock Pattern",
          "pattern_id": "mock-pattern-id",
          "stops": Array [
            Object {
              "stop_id": "1",
              "stop_lat": 36.9707,
              "stop_lon": -122.0246,
              "stop_name": "Santa Cruz",
            },
            Object {
              "stop_id": "2",
              "stop_lat": 37.049,
              "stop_lon": -122.0279,
              "stop_name": "Scotts Valley",
            },
          ],
          "trips": Array [
            Object {
              "duration": 1234,
              "end_time": 13579,
              "start_time": 12345,
              "trip_headsign": "To Downtown",
              "trip_id": "abcd",
              "trip_short_name": "The Express",
            },
          ],
        },
      ]
    }
  >
    <Styled(div)
      __css={
        Object {
          "position": "relative",
          "width": "100%",
        }
      }
      className="chakra-form-control"
      role="group"
      routePatterns={
        Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                Array [
                  -122.0246,
                  36.9707,
                ],
                Array [
                  -122.0279,
                  37.049,
                ],
                Array [
                  -121.9799,
                  37.2299,
                ],
                Array [
                  -121.9445,
                  37.324,
                ],
                Array [
                  -121.936,
                  37.353,
                ],
                Array [
                  -121.924,
                  37.365,
                ],
              ],
              "type": "LineString",
            },
            "name": "Mock Pattern",
            "pattern_id": "mock-pattern-id",
            "stops": Array [
              Object {
                "stop_id": "1",
                "stop_lat": 36.9707,
                "stop_lon": -122.0246,
                "stop_name": "Santa Cruz",
              },
              Object {
                "stop_id": "2",
                "stop_lat": 37.049,
                "stop_lon": -122.0279,
                "stop_name": "Scotts Valley",
              },
            ],
            "trips": Array [
              Object {
                "duration": 1234,
                "end_time": 13579,
                "start_time": 12345,
                "trip_headsign": "To Downtown",
                "trip_id": "abcd",
                "trip_short_name": "The Express",
              },
            ],
          },
        ]
      }
    >
      <div
        className="chakra-form-control css-1kxonj9"
        role="group"
        routePatterns={
          Array [
            Object {
              "geometry": Object {
                "coordinates": Array [
                  Array [
                    -122.0246,
                    36.9707,
                  ],
                  Array [
                    -122.0279,
                    37.049,
                  ],
                  Array [
                    -121.9799,
                    37.2299,
                  ],
                  Array [
                    -121.9445,
                    37.324,
                  ],
                  Array [
                    -121.936,
                    37.353,
                  ],
                  Array [
                    -121.924,
                    37.365,
                  ],
                ],
                "type": "LineString",
              },
              "name": "Mock Pattern",
              "pattern_id": "mock-pattern-id",
              "stops": Array [
                Object {
                  "stop_id": "1",
                  "stop_lat": 36.9707,
                  "stop_lon": -122.0246,
                  "stop_name": "Santa Cruz",
                },
                Object {
                  "stop_id": "2",
                  "stop_lat": 37.049,
                  "stop_lon": -122.0279,
                  "stop_name": "Scotts Valley",
                },
              ],
              "trips": Array [
                Object {
                  "duration": 1234,
                  "end_time": 13579,
                  "start_time": 12345,
                  "trip_headsign": "To Downtown",
                  "trip_id": "abcd",
                  "trip_short_name": "The Express",
                },
              ],
            },
          ]
        }
      >
        <FormLabel
          htmlFor="Patterns"
        >
          <Styled(label)
            __css={
              Object {
                "_disabled": Object {
                  "opacity": 0.4,
                },
                "display": "block",
                "fontSize": "md",
                "fontWeight": "medium",
                "marginEnd": 3,
                "mb": 2,
                "opacity": 1,
                "textAlign": "start",
                "transition": "all 0.2s",
              }
            }
            className="chakra-form__label"
            htmlFor="Patterns"
            id="field-1-label"
          >
            <label
              className="chakra-form__label css-10m1y3x"
              htmlFor="Patterns"
              id="field-1-label"
            >
              Select patterns
            </label>
          </Styled(label)>
        </FormLabel>
        <ForwardRef
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          inputId="Patterns"
          isMulti={true}
          name="Patterns"
          onChange={[Function]}
          options={
            Array [
              Object {
                "geometry": Object {
                  "coordinates": Array [
                    Array [
                      -122.0246,
                      36.9707,
                    ],
                    Array [
                      -122.0279,
                      37.049,
                    ],
                    Array [
                      -121.9799,
                      37.2299,
                    ],
                    Array [
                      -121.9445,
                      37.324,
                    ],
                    Array [
                      -121.936,
                      37.353,
                    ],
                    Array [
                      -121.924,
                      37.365,
                    ],
                  ],
                  "type": "LineString",
                },
                "name": "Mock Pattern",
                "pattern_id": "mock-pattern-id",
                "stops": Array [
                  Object {
                    "stop_id": "1",
                    "stop_lat": 36.9707,
                    "stop_lon": -122.0246,
                    "stop_name": "Santa Cruz",
                  },
                  Object {
                    "stop_id": "2",
                    "stop_lat": 37.049,
                    "stop_lon": -122.0279,
                    "stop_name": "Scotts Valley",
                  },
                ],
                "trips": Array [
                  Object {
                    "duration": 1234,
                    "end_time": 13579,
                    "start_time": 12345,
                    "trip_headsign": "To Downtown",
                    "trip_id": "abcd",
                    "trip_short_name": "The Express",
                  },
                ],
              },
            ]
          }
          placeholder="Select patterns"
          value={
            Array [
              Object {
                "geometry": Object {
                  "coordinates": Array [
                    Array [
                      -122.0246,
                      36.9707,
                    ],
                    Array [
                      -122.0279,
                      37.049,
                    ],
                    Array [
                      -121.9799,
                      37.2299,
                    ],
                    Array [
                      -121.9445,
                      37.324,
                    ],
                    Array [
                      -121.936,
                      37.353,
                    ],
                    Array [
                      -121.924,
                      37.365,
                    ],
                  ],
                  "type": "LineString",
                },
                "name": "Mock Pattern",
                "pattern_id": "mock-pattern-id",
                "stops": Array [
                  Object {
                    "stop_id": "1",
                    "stop_lat": 36.9707,
                    "stop_lon": -122.0246,
                    "stop_name": "Santa Cruz",
                  },
                  Object {
                    "stop_id": "2",
                    "stop_lat": 37.049,
                    "stop_lon": -122.0279,
                    "stop_name": "Scotts Valley",
                  },
                ],
                "trips": Array [
                  Object {
                    "duration": 1234,
                    "end_time": 13579,
                    "start_time": 12345,
                    "trip_headsign": "To Downtown",
                    "trip_id": "abcd",
                    "trip_short_name": "The Express",
                  },
                ],
              },
            ]
          }
        >
          <ReactSelect
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            innerRef={null}
            inputId="Patterns"
            isMulti={true}
            name="Patterns"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "geometry": Object {
                    "coordinates": Array [
                      Array [
                        -122.0246,
                        36.9707,
                      ],
                      Array [
                        -122.0279,
                        37.049,
                      ],
                      Array [
                        -121.9799,
                        37.2299,
                      ],
                      Array [
                        -121.9445,
                        37.324,
                      ],
                      Array [
                        -121.936,
                        37.353,
                      ],
                      Array [
                        -121.924,
                        37.365,
                      ],
                    ],
                    "type": "LineString",
                  },
                  "name": "Mock Pattern",
                  "pattern_id": "mock-pattern-id",
                  "stops": Array [
                    Object {
                      "stop_id": "1",
                      "stop_lat": 36.9707,
                      "stop_lon": -122.0246,
                      "stop_name": "Santa Cruz",
                    },
                    Object {
                      "stop_id": "2",
                      "stop_lat": 37.049,
                      "stop_lon": -122.0279,
                      "stop_name": "Scotts Valley",
                    },
                  ],
                  "trips": Array [
                    Object {
                      "duration": 1234,
                      "end_time": 13579,
                      "start_time": 12345,
                      "trip_headsign": "To Downtown",
                      "trip_id": "abcd",
                      "trip_short_name": "The Express",
                    },
                  ],
                },
              ]
            }
            placeholder="Select patterns"
            styles={
              Object {
                "control": [Function],
                "menu": [Function],
                "option": [Function],
              }
            }
            value={
              Array [
                Object {
                  "geometry": Object {
                    "coordinates": Array [
                      Array [
                        -122.0246,
                        36.9707,
                      ],
                      Array [
                        -122.0279,
                        37.049,
                      ],
                      Array [
                        -121.9799,
                        37.2299,
                      ],
                      Array [
                        -121.9445,
                        37.324,
                      ],
                      Array [
                        -121.936,
                        37.353,
                      ],
                      Array [
                        -121.924,
                        37.365,
                      ],
                    ],
                    "type": "LineString",
                  },
                  "name": "Mock Pattern",
                  "pattern_id": "mock-pattern-id",
                  "stops": Array [
                    Object {
                      "stop_id": "1",
                      "stop_lat": 36.9707,
                      "stop_lon": -122.0246,
                      "stop_name": "Santa Cruz",
                    },
                    Object {
                      "stop_id": "2",
                      "stop_lat": 37.049,
                      "stop_lon": -122.0279,
                      "stop_name": "Scotts Valley",
                    },
                  ],
                  "trips": Array [
                    Object {
                      "duration": 1234,
                      "end_time": 13579,
                      "start_time": 12345,
                      "trip_headsign": "To Downtown",
                      "trip_id": "abcd",
                      "trip_short_name": "The Express",
                    },
                  ],
                },
              ]
            }
          >
            <div
              className="ReactSelectMock"
            />
          </ReactSelect>
        </ForwardRef>
      </div>
    </Styled(div)>
  </FormControl>
</SelectPatterns>
`;
